documentation/en/dev/architecture.md

```diff
@@ -0,0 +1,298 @@
+
+The two most important stages of the sync are:
+
+* `StageHeaders`: collect all the blocks (through the `BlockSync`) needed to connect our chain with the new head. The *header* synecdoche, used throughout the code, refers to the actual Filecoin blocks, which are identified by just their headers (`BlockHeader`) until we actually need the messages associated to it (converting it to a `FullBlock`), but normally we can think of them as synonyms.
+
+* `StageMessages`: once we have all the blocks connecting our chain to the new head we can update the chain state up to the head by sequentially applying all the messages associated with the collected blocks. That is the *only* way to verify if a head is valid, we need to evaluate its state, and for that we need to execute all the messages necessary to arrive to it (see VM section later for a full discussion about state).
+
+The two functions associated with these phases are `(*Syncer).collectHeaders()` and `syncMessagesAndCheckState()` respectively. Most of the state execution and validation of the second one happens inside `(*Syncer).ValidateBlock()` (a few calls down the line, as we need first to split a tipset into its constituent blocks and analyze one at a time).
+
+## Setting the new head in the chain
+
+If all validations pass we will now set that head as our heaviest tipset and, since all the collection and validation already happened as a side effect of the sync process, we will already have the chain in the synced state to reflect it.
+
+In `Sync()`, after `collectChain()` is done (this function does more than its name implies) we will have synced to the new head, the sync state which will be reflected in the CLI command is `StageSyncComplete`. The next function called there is `(*ChainStore).PutTipSet()`, eventually leading to `takeHeaviestTipSet()` that will update the `heaviest` pointer in the `ChainStore` to put to the new head (tipset) received.
+
+It is important to note at this point that similar to the IPFS architecture of addressing by content and not by location/address (FIXME: check and link to IPFS docs) the "actual" chain stored in the node repo is *relative* to which CID we look for. We always have stored a series of Filecoin blocks pointing to other blocks, each a potential chain in itself by following its parent's reference, and its parent's parent, and so on up to the genesis block. (FIXME: We need a diagram here, one of the Filecoin blog entries might have something similar to what we are describing here.) It only depends on *where* (location) do we start to look for. The *only* address/location reference we hold of the chain, a relative reference, is the `heaviest` pointer. This is reflected by the fact that we don't store it in the `Blockstore` by a fixed, *absolute*, CID that reflects its contents, as this will change each time we sync to a new head (FIXME: link to the immutability IPFS doc that I need to write).
```

Comment:
Mutation of state in IPFS: https://github.com/ipfs/ipfs-docs/issues/264

(For now let's just point to the issue with my notes until someone from IPFS docs can polish it and merge it into the site.)